### PR TITLE
[MRESOLVER-283] Shared executor service

### PR DESCRIPTION
**Pull request by cstamas**
More and more component in resolver does parallel processing (BF collector, MD resolver, basic connector), and they all create, maintain their own executor instance.

Instead of this, create one shared service component and just reuse it accross resolver.

---

https://issues.apache.org/jira/browse/MRESOLVER-283

### CODE REVIEW

**Comment by cstamas** [comment]
@caiwei-ebay ping

**Comment by caiwei-ebay** [comment]
@cstamas 

The PR looks good to me. Thanks for the quick fix.
My main concern is about how to tell user pom/metadata.xml is downloaded in parallel now with BF collector.

As the special param "aether.dependencyCollector.bf.threads" has to be removed using SharedExecutor, could you put some lines to explain the parallel download behavior in param "aether.dependencyCollector.impl" where you've introduced BF and DF.  With such information, user may try the BF and feedback so we can discuss whether BF could be the default based on the feedbacks.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutor.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+                new LinkedBlockingQueue<>(), new WorkerThreadFactory( getClass().getSimpleName() + '-' ) );
+    }
+
+    @Inject
+    public DefaultResolverExecutor( @Named( "${" + CONF_PROP_THREADS + ":-" + DEFAULT_THREADS + "}" ) final int threads,
+                                    RepositorySystemLifecycle lifecycle )
+    {
+        if ( threads < 1 )
+        {
+            throw new IllegalArgumentException( "threads must be grater than zero" );
```
Comment (caiwei-ebay) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutor.java:75]: typo grator -> greator

File: maven-resolver-connector-basic/src/main/java/org/eclipse/aether/connector/basic/BasicRepositoryConnector.java [hunk 2]
```diff
@@ -656,18 +622,4 @@ private void uploadChecksum( URI location, Object checksum )
...
-    {
-
-        static final Executor INSTANCE = new DirectExecutor();
-
-        @Override
-        public void execute( Runnable command )
-        {
-            command.run();
-        }
-
```
Comment (caiwei-ebay) [inline maven-resolver-connector-basic/src/main/java/org/eclipse/aether/connector/basic/BasicRepositoryConnector.java:671]: Given this being deleted, I think it may introduce additional cost for DF.
DF downloads pom one by one, originally it will be downloaded by main thread directly and now it will be downloaded in a thread and the main thread has to wait

File: src/site/markdown/configuration.md [hunk 3]
```diff
@@ -52,7 +52,6 @@ Option | Type | Description | Default Value | Supports Repo ID Suffix
 `aether.dependencyCollector.maxExceptions` | int | Only exceptions up to the number given in this configuration property are emitted. Exceptions which exceed that number are swallowed. | `50` | no
 `aether.dependencyCollector.impl` | String | The name of the dependency collector implementation to use: depth-first (original) named `df`, and breadth-first (new in 1.8.0) named `bf`. Both collectors produce equivalent results, but they may differ performance wise, depending on project being applied to. Our experience shows that existing `df` is well suited for smaller to medium size projects, while `bf` may perform better on huge projects with many dependencies. Experiment (and come back to us!) to figure out which one suits you the better. | `"df"` | no
```
Comment (caiwei-ebay) [inline src/site/markdown/configuration.md:53]: Shall we add more info about "BF will download poms/metadata.xml in parallel"?

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/bf/BfDependencyCollector.java [hunk 4]
```diff
@@ -475,23 +482,22 @@ else if ( descriptorResult == DataPool.NO_DESCRIPTOR )
...
         {
-            this.executorService = getExecutorService( session );
+            this.executor = executor;
         }
 
         void resolveDescriptors( Artifact artifact, Callable<DescriptorResolutionResult> callable )
         {
             results.computeIfAbsent( ArtifactIdUtils.toId( artifact ),
-                    key -> this.executorService.submit( callable ) );
+                    key -> this.executor.submit( callable ) );
```
Comment (caiwei-ebay) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/bf/BfDependencyCollector.java:500]: In the illustration below:

https://camo.githubusercontent.com/71a9619274e478245df1c37423239b79f6b3036a3af72690d7ace08419ee21ac/68747470733a2f2f6973737565732e6170616368652e6f72672f6a6972612f7365637572652f6174746163686d656e742f31333034373931392f7265736f6c76655f646570732e706e67

As we submit a Callable<DescriptorResolutionResult> here, and this callable finally resorts to RepositoryConnector which will then submit a Collection<Artifact> (a collection with single artifact) to the executor, 2 different tasks running in one executor and besides the 2 tasks has dependencies.

The original behavior is:

we submit a Callable<DescriptorResolutionResult> here, and this callable finally resorts to RepositoryConnector which will then submit a Collection<Artifact> (a collection with single artifact) to the executor, but this time the executor is actually a DirectExecutor which actually runs in main thread. Please check code below.

if ( maxThreads <= 1 )
        {
            return DirectExecutor.INSTANCE;
        }
        int tasks = safe( artifacts ).size() + safe( metadatas ).size();
        if ( tasks <= 1 )
        {
            return DirectExecutor.INSTANCE;
        }

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutor.java (hunk 1 above)
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutor.java]: fixed

File: maven-resolver-connector-basic/src/main/java/org/eclipse/aether/connector/basic/BasicRepositoryConnector.java (hunk 2 above)
Comment (cstamas) [reply maven-resolver-connector-basic/src/main/java/org/eclipse/aether/connector/basic/BasicRepositoryConnector.java:671]: reintroduced direct strategy as well, two out of 3 users of resolver executor now relies on it.

File: src/site/markdown/configuration.md (hunk 3 above)
Comment (cstamas) [reply src/site/markdown/configuration.md:52]: I think a dedicated page would be better, that on this, already very dense configuration page.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/bf/BfDependencyCollector.java (hunk 4 above)
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/bf/BfDependencyCollector.java:500]: Hopefully I restored the original behaviour, please recheck (latest commit)

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutor.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    @Override
+    public void initService( ServiceLocator locator )
+    {
+        locator.getService( RepositorySystemLifecycle.class ).addOnSystemEndedHandler( this::shutdown );
+    }
+
+    @Override
+    public void submitOrDirect( Collection<Runnable> tasks )
+    {
+        requireNonNull( tasks );
```
Comment (caiwei-ebay) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutor.java:111]: When resolves a snapshot dependency or dependency with version range , Maven resolves metadata.xml from multiple repositories in parallel: https://github.com/apache/maven/blob/master/maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java#L158

Suppose we have multiple repositories like releases, snapshots, commercial, legacy, testing in our private Maven nexus repository and we need to resolve 5 dependencies (snapshot or with version range) in parallel. 

When resolve with BFDependencyCollector,  5 tasks to resolve each dependency have used up all 5 threads, and due to resolving each dependency (task A) need to resolve metadata (task B), the metadata resolution task B will be also submitted to the same executor. A depends on B but B has to wait available threads. I think this could be lead to endless waiting... No?

As it is better to use separate thread pool for separate tasks, maybe the SharedExecutor here should have separate pool for separate tasks? At least, all poms downloading used one pool and all metadata.xml downloading used another pool. Please advice.



-

**Comment by cstamas** [comment]
Reworked to:
* have split pools (service can decide using service "key")
* pools are stored in session, so each session has own pool(s)
* this merely reflects what was before, but all is centralized.

**Review by caiwei-ebay — APPROVED** [review]
The PR looks great to me. Thanks for tidy up the design.

**Comment by michael-o** [comment]
Will try to review today

File: maven-resolver-connector-basic/src/main/java/org/eclipse/aether/connector/basic/BasicRepositoryConnector.java [hunk 6]
```diff
@@ -225,11 +205,13 @@ public void get( Collection<? extends ArtifactDownload> artifactDownloads,
             throw new IllegalStateException( "connector closed" );
         }
 
-        Executor executor = getExecutor( artifactDownloads, metadataDownloads );
         RunnableErrorForwarder errorForwarder = new RunnableErrorForwarder();
         List<ChecksumAlgorithmFactory> checksumAlgorithmFactories = layout.getChecksumAlgorithmFactories();
+        Collection<? extends MetadataDownload> mds = safe( metadataDownloads );
+        Collection<? extends ArtifactDownload> ads = safe( artifactDownloads );
+        ArrayList<Runnable> runnable = new ArrayList<>( mds.size() + ads.size() );
```
Comment (michael-o) [inline maven-resolver-connector-basic/src/main/java/org/eclipse/aether/connector/basic/BasicRepositoryConnector.java:212]: runnables

File: maven-resolver-connector-basic/src/main/java/org/eclipse/aether/connector/basic/BasicRepositoryConnector.java [hunk 7]
```diff
@@ -146,8 +150,10 @@
         this.repository = repository;
         this.fileProcessor = fileProcessor;
         this.providedChecksumsSources = providedChecksumsSources;
+        this.resolverExecutor = resolverExecutorService.getResolverExecutor( session, RepositoryConnector.class,
+                ConfigUtils.getInteger( session, CONFIG_PROP_THREADS_DEFAULT, CONFIG_PROP_THREADS,
+                        "maven.artifact.threads" ) );
```
Comment (michael-o) [inline maven-resolver-connector-basic/src/main/java/org/eclipse/aether/connector/basic/BasicRepositoryConnector.java:155]: We should also deprecate this property since it does not reflect reality. MD is not artifacts, but this connector does both.

File: maven-resolver-connector-basic/src/main/java/org/eclipse/aether/connector/basic/BasicRepositoryConnector.java [hunk 8]
```diff
@@ -287,9 +269,10 @@ public void get( Collection<? extends ArtifactDownload> artifactDownloads,
                 task = new GetTaskRunner( location, transfer.getFile(), checksumPolicy,
                         checksumAlgorithmFactories, checksumLocations, providedChecksums, listener );
             }
-            executor.execute( errorForwarder.wrap( task ) );
+            runnable.add( errorForwarder.wrap( task ) );
         }
 
+        resolverExecutor.submitOrDirect( runnable );
```
Comment (michael-o) [inline maven-resolver-connector-basic/src/main/java/org/eclipse/aether/connector/basic/BasicRepositoryConnector.java:275]: Why wait until they are all collected instead separate MD and A like before?

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultMetadataResolver.java [hunk 9]
```diff
@@ -83,8 +78,16 @@
...
 
+    /**
+     * The count of threads to be used when resolving metadata in parallel, default value 4.
+     */
     private static final String CONFIG_PROP_THREADS = "aether.metadataResolver.threads";
 
+    /**
+     * The default value for {@link #CONFIG_PROP_THREADS}.
+     */
+    private static final int CONFIG_PROP_THREADS_DEFAULT = 4;
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultMetadataResolver.java:89]: Why not 5 like in the other one?

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultMetadataResolver.java [hunk 10]
```diff
@@ -374,41 +391,36 @@ else if ( exception == null )
 
         if ( !tasks.isEmpty() )
         {
-            int threads = ConfigUtils.getInteger( session, 4, CONFIG_PROP_THREADS );
-            Executor executor = getExecutor( Math.min( tasks.size(), threads ) );
-            try
+            RunnableErrorForwarder errorForwarder = new RunnableErrorForwarder();
+            ArrayList<Runnable> runnable = new ArrayList<>( tasks.size() );
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultMetadataResolver.java:395]: runnables

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutor.java [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+{
+    private final ExecutorService executorService;
+
+    DefaultResolverExecutor( final ExecutorService executorService )
+    {
+        this.executorService = executorService;
+    }
+
+    @Override
+    public void submitOrDirect( Collection<Runnable> tasks )
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutor.java:49]: I think from an abstract PoV just `submit` is enough. It is an implementation detail whether it should be executed directly or scheduled for execution. The behavior should be described in the Javadoc of the class.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutorService.java [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    @Override
+    public ResolverExecutor getResolverExecutor( RepositorySystemSession session,
+                                                 Class<?> service,
+                                                 int maxThreads )
+    {
+        requireNonNull( session );
+        requireNonNull( service );
+        if ( maxThreads < 1 )
+        {
+            throw new IllegalArgumentException( "threads must be greater than zero" );
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutorService.java:58]: maxThreads must be...

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutorService.java [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+    private ExecutorService createExecutorService( Class<?> service, int maxThreads )
+    {
+        ThreadPoolExecutor threadPoolExecutor = new ThreadPoolExecutor(
+                maxThreads,
+                maxThreads,
+                3L, TimeUnit.SECONDS,
+                new LinkedBlockingQueue<>(),
+                new WorkerThreadFactory( getClass().getSimpleName() + "-" + service.getSimpleName() + "-" )
+        );
+        threadPoolExecutor.allowCoreThreadTimeOut( true );
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutorService.java:89]: Why is this necessary?

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutorService.java [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+     * currently there is no way to shut down them.
+     */
+    private ExecutorService createExecutorService( Class<?> service, int maxThreads )
+    {
+        ThreadPoolExecutor threadPoolExecutor = new ThreadPoolExecutor(
+                maxThreads,
+                maxThreads,
+                3L, TimeUnit.SECONDS,
+                new LinkedBlockingQueue<>(),
+                new WorkerThreadFactory( getClass().getSimpleName() + "-" + service.getSimpleName() + "-" )
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutorService.java:87]: So we have lost the per repo thread pool?

**Review by michael-o — CHANGES_REQUESTED** [review]

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutor.java (hunk 5 above)
Comment (michael-o) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutor.java:51]: I'd like also add as I understand the artifact thread pools where per repo, one this is one per all repos?

I think we need a overview what executor services where created previously and which now to compare whether this could create a bottleneck.

File: maven-resolver-connector-basic/src/main/java/org/eclipse/aether/connector/basic/BasicRepositoryConnector.java (hunk 7 above)
Comment (cstamas) [reply maven-resolver-connector-basic/src/main/java/org/eclipse/aether/connector/basic/BasicRepositoryConnector.java:155]: Maybe we should just remove it? We are going for 1.9 after all....

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultMetadataResolver.java (hunk 9 above)
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultMetadataResolver.java:89]: I kept same values as before change, this component used by default 4 threads, no change in that area.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutor.java (hunk 11 above)
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutor.java:49]: I wanted to clearly distinquish, that's all, moreover, if `submit` then we overload the other method as well.... but right, semantics should not be exposed, so maybe `submitBatch`?

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutorService.java (hunk 13 above)
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutorService.java:89]: To let the pool become garbage collected, otherwise (if have living threads) it will NOT be GC-ed. OTOH, for performance reasons, we do want to create pool (and keep the pool) with live threads for some....

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutorService.java (hunk 14 above)
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutorService.java:87]: Good point, fixing

File: maven-resolver-connector-basic/src/main/java/org/eclipse/aether/connector/basic/BasicRepositoryConnector.java (hunk 8 above)
Comment (cstamas) [reply maven-resolver-connector-basic/src/main/java/org/eclipse/aether/connector/basic/BasicRepositoryConnector.java:275]: What is the difference? It's not this line that wait but next, the unchanged one.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutor.java (hunk 11 above)
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutor.java]: renamed to `submitBatch`

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutorService.java (hunk 12 above)
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutorService.java]: Fixed

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutorService.java (hunk 14 above)
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutorService.java]: Fixed

**Review by gnodet — CHANGES_REQUESTED** [review]
Sharing executors must be done with caution.

Is there any way that the executor may be running out of threads to service a call ? For example, resolving multiple files in parallel, each file being resolved in its own thread, then downloading dependencies and running out of threads ?

I think using a more general `ForkJoinPool` which supports work stealing would be safer to avoid running into such problems.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutor.java (hunk 5 above)
Comment (gnodet) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/concurrency/DefaultResolverExecutor.java:51]: Bottleneck... or deadlock caused by thread exhaustion.

**Comment by cstamas** [comment]
> Sharing executors must be done with caution.
> 
> Is there any way that the executor may be running out of threads to service a call ? For example, resolving multiple files in parallel, each file being resolved in its own thread, then downloading dependencies and running out of threads ?
> 
> I think using a more general `ForkJoinPool` which supports work stealing would be safer to avoid running into such problems.

I kinda agree with @gnodet  here, and will "tone down" this PR to NOT SHARE executors, merely will introduce this facade but instances will be handled as before (as even now, it is not 100% what it was before: collector had per arg instance, now shared). And I cannot certainly asnwer the question "Is there any way that the executor may be running out of threads to service a call".

OTOH, I disagree with forkjoinpool: all these threads will do (probably blocking, maybe inifinite) HTTP IO, not to mention the refactoring needed to use them (most of "client" code remained unchanged here, using plain Runnable and Callable).

---

Hence, this PR will change to merely "centralize executor creation and (probably) use ResolverExecutor shim instead directly exposing ExecutorService iface"...

**Comment by gnodet** [comment]
> OTOH, I disagree with forkjoinpool: all these threads will do (probably blocking, maybe inifinite) HTTP IO, not to mention the refactoring needed to use them (most of "client" code remained unchanged here, using plain Runnable and Callable).

`ForkJoinPool` is capable of handling blocking operations, just like a standard `ThreadPoolExecutor`. The only problem is when you don't use your own instance and reuse the system provided instance with blocking operations.  The only difference between a `ForkJoinPool` and a `ThreadPoolExecutor` is that the former has more features such as thread affinity, work stealing and fork/join operations.

I think a single thread pool is a good idea, but it has to be implemented in a correct way to avoid any thread exhaustion. I.e. use a single fork/join pool and use the appropriate methods to dispatch all the work, which should ensure proper ordering of tasks execution.  I've done a similar thing in https://github.com/apache/maven/pull/803 (which I need to finish).

**Comment by cstamas** [comment]
Ok, now I feel I should not push this anymore, as I agree now with use of forkJoinPool. Hence, will close this PR, as that change would be bigger, but I want 1.9.0 out as soon as possible....

**Comment by jira-importer** [comment]
Resolve #959
